app/grandchallenge/evaluation/models.py

```diff
@@ -140,6 +140,70 @@
 }
 
 
+def get_archive_items_for_interfaces(*, algorithm_interfaces, archive_items):
+    valid_archive_items_per_interface = {}
+    for interface in algorithm_interfaces:
+        inputs = interface.inputs.all()
+        valid_archive_items_per_interface[interface] = archive_items.annotate(
+            input_count=Count("values", distinct=True),
+            relevant_input_count=Count(
+                "values", filter=Q(values__interface__in=inputs), distinct=True
+            ),
+        ).filter(input_count=len(inputs), relevant_input_count=len(inputs))
+    return valid_archive_items_per_interface
+
+
+def get_valid_jobs_for_interfaces_and_archive_items(
+    *,
+    algorithm_interfaces,
+    jobs,
+    valid_archive_items_per_interface,
+):
+    # get jobs whose inputs match those of the given algorithm interfaces
+    # regardless of job status
+
+    jobs_per_interface = {}
+    for interface in algorithm_interfaces:
+        jobs_per_interface[interface] = []
+        input_count = interface.inputs.count()
+
+        interface_jobs = (
+            jobs.filter(algorithm_interface=interface)
+            .annotate(
+                input_count=Count("inputs", distinct=True),
+                input_interface_match_count=Count(
+                    "inputs",
+                    filter=Q(inputs__interface__in=interface.inputs.all()),
+                    distinct=True,
+                ),
+            )
+            .filter(
+                # first filter out jobs whose inputs don't match its interface,
+                # this can happen because we migrated assuming each job for
+                # an algorithm was done with the latest sets of inputs/outputs
+                input_count=input_count,
+                input_interface_match_count=input_count,
+            )
```

Comment:
This might not be necessary. It's only relevant as long as there are jobs whose inputs don't match their interface. For jobs created from now on, this should never happen, so as time goes by we will run into this less and less. For now, it reduces the number of jobs to iterate over in the for-loop below and hence might help with speed, but the annotation also costs time of course, possibly resulting in net 0 difference. We should benchmark this. I will put together some queries for comparison.

Reply:
This is no more expensive than the annotation we currently do, so needs no extra benchmarking I think.